File: pgpainless-core/src/main/java/org/pgpainless/util/selection/userid/SelectUserId.java
// SPDX-FileCopyrightText: 2021 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.util.selection.userid;

import java.util.ArrayList;
import java.util.List;

import org.bouncycastle.openpgp.PGPKeyRing;
import org.pgpainless.PGPainless;
import org.pgpainless.key.info.KeyRingInfo;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * Filter for selecting user-ids from keys and from lists.
 */
public abstract class SelectUserId {

    /**
     * Return true, if the given user-id is accepted by this particular filter, false otherwise.
     *
     * @param userId user-id
     * @return acceptance of the filter
     */
    protected abstract boolean accept(String userId);

    /**
     * Select all currently valid user-ids of the given key ring.
     *
     * @param keyRing public or secret key ring
     * @return valid user-ids
     */
    @Nonnull
    public List<String> selectUserIds(@Nonnull PGPKeyRing keyRing) {
        List<String> userIds = PGPainless.inspectKeyRing(keyRing).getValidUserIds();
        return selectUserIds(userIds);
    }

    /**
     * Select all acceptable (see {@link #accept(String)}) from the given list of user-ids.
     *
     * @param userIds list of user-ids
     * @return sub-list of acceptable user-ids
     */
    @Nonnull
    public List<String> selectUserIds(@Nonnull List<String> userIds) {
        List<String> selected = new ArrayList<>();
        for (String userId : userIds) {
            if (accept(userId)) {
                selected.add(userId);
            }
        }
        return selected;
    }

    /**
     * Return the first valid, acceptable user-id from the given public or secret key ring.
     *
     * @param keyRing public or secret key ring
     * @return first matching valid user-id or null
     */
    @Nullable
    public String firstMatch(PGPKeyRing keyRing) {
        return firstMatch(selectUserIds(keyRing));
    }

    /**
     * Return the first valid, acceptable user-id from the list of user-ids.
     *
     * @param userIds list of user-ids
     * @return first matching valid user-id or null
     */
    @Nullable
    public String firstMatch(@Nonnull List<String> userIds) {
        for (String userId : userIds) {
            if (accept(userId)) {
                return userId;
            }
        }
        return null;
    }

    /**
     * Filter that filters for user-ids which contain the given <pre>query</pre> as a substring.
     *
     * @param query query
     * @return filter
     */
    public static SelectUserId containsSubstring(@Nonnull CharSequence query) {
        return new SelectUserId() {
            @Override
            protected boolean accept(String userId) {
                return userId.contains(query.toString());
            }
        };
    }

    /**
     * Filter that filters for user-ids which match the given <pre>query</pre> exactly.
     *
     * @param query query
     * @return filter
     */
    public static SelectUserId exactMatch(@Nonnull CharSequence query) {
        return new SelectUserId() {
            @Override
            protected boolean accept(String userId) {
                return userId.equals(query.toString());
            }
        };
    }

    /**
     * Filter that filters for user-ids which start with the given <pre>substring</pre>.
     *
     * @param substring substring
     * @return filter
     */
    public static SelectUserId startsWith(@Nonnull CharSequence substring) {
        String string = substring.toString();
        return new SelectUserId() {
            @Override
            protected boolean accept(String userId) {
                return userId.startsWith(string);
            }
        };
    }

    /**
     * Filter that filters for user-ids which contain the given <pre>email</pre> address.
     * Note: This only accepts user-ids which properly have the email address surrounded by angle brackets.
     *
     * The argument <pre>email</pre> can both be a plain email address (<pre>"foo@bar.baz"</pre>),
     * or surrounded by angle brackets (<pre>"<foo@bar.baz>"</pre>, the result of the filter will be the same.
     *
     * @param email email address
     * @return filter
     */
    public static SelectUserId containsEmailAddress(@Nonnull CharSequence email) {
        String string = email.toString();
        return containsSubstring(string.matches("^<.+>$") ? string : '<' + string + '>');
    }

    /**
     * Filter that filters for valid user-ids on the given <pre>keyRing</pre> only.
     *
     * @param keyRing public / secret keys
     * @return filter
     */
    public static SelectUserId validUserId(PGPKeyRing keyRing) {
        final KeyRingInfo info = PGPainless.inspectKeyRing(keyRing);

        return new SelectUserId() {
            @Override
            protected boolean accept(String userId) {
                return info.isUserIdValid(userId);
            }
        };
    }

    /**
     * Filter that filters for user-ids which pass all the given <pre>filters</pre>.
     *
     * @param filters filters
     * @return filter
     */
    public static SelectUserId and(SelectUserId... filters) {
        return new SelectUserId() {
            @Override
            protected boolean accept(String userId) {
                boolean accept = true;
                for (SelectUserId filter : filters) {
                    accept &= filter.accept(userId);
                }
                return accept;
            }
        };
    }

    /**
     * Filter that filters for user-ids which pass at least one of the given <pre>filters</pre>.
     *
     * @param filters filters
     * @return filter
     */
    public static SelectUserId or(SelectUserId... filters) {
        return new SelectUserId() {
            @Override
            protected boolean accept(String userId) {
                boolean accept = false;
                for (SelectUserId filter : filters) {
                    accept |= filter.accept(userId);
                }
                return accept;
            }
        };
    }

    /**
     * Filter that inverts the result of the given <pre>filter</pre>.
     *
     * @param filter filter
     * @return inverting filter
     */
    public static SelectUserId not(SelectUserId filter) {
        return new SelectUserId() {
            @Override
            protected boolean accept(String userId) {
                return !filter.accept(userId);
            }
        };
    }

    /**
     * Filter that selects user-ids by the given <pre>email</pre> address.
     * It returns user-ids which either contain the given <pre>email</pre> address as angle-bracketed string,
     * or which equal the given <pre>email</pre> string exactly.
     *
     * @param email email
     * @return filter
     */
    public static SelectUserId byEmail(CharSequence email) {
        return SelectUserId.or(
                SelectUserId.exactMatch(email),
                SelectUserId.containsEmailAddress(email)
        );
    }
}

File: pgpainless-core/src/main/kotlin/org/pgpainless/util/selection/userid/SelectUserId.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.util.selection.userid

import java.util.function.Predicate
import org.bouncycastle.openpgp.PGPKeyRing
import org.bouncycastle.openpgp.api.OpenPGPCertificate
import org.pgpainless.PGPainless

abstract class SelectUserId : Predicate<String>, (String) -> Boolean {

    /** Legacy glue code to forward accept() calls to invoke() instead. */
    @Deprecated("Use invoke() instead.", ReplaceWith("invoke(userId)"))
    protected fun accept(userId: String): Boolean = invoke(userId)

    override fun test(userId: String): Boolean = invoke(userId)

    companion object {

        /**
         * Filter for user-ids which match the given [query] exactly.
         *
         * @param query query
         * @return filter
         */
        @JvmStatic
        fun exactMatch(query: CharSequence) =
            object : SelectUserId() {
                override fun invoke(userId: String): Boolean = userId == query
            }

        /**
         * Filter for user-ids which start with the given [substring].
         *
         * @param substring substring
         * @return filter
         */
        @JvmStatic
        fun startsWith(substring: CharSequence) =
            object : SelectUserId() {
                override fun invoke(userId: String): Boolean = userId.startsWith(substring)
            }

        /**
         * Filter for user-ids which contain the given [substring].
         *
         * @param substring query
         * @return filter
         */
        @JvmStatic
        fun containsSubstring(substring: CharSequence) =
            object : SelectUserId() {
                override fun invoke(userId: String): Boolean = userId.contains(substring)
            }

        /**
         * Filter for user-ids which contain the given [email] address. Note: This only accepts
         * user-ids which properly have the email address surrounded by angle brackets.
         *
         * The argument [email] can both be a plain email address (`foo@bar.baz`), or surrounded by
         * angle brackets (`<foo@bar.baz>`), the result of the filter will be the same.
         *
         * @param email email address
         * @return filter
         */
        @JvmStatic
        fun containsEmailAddress(email: CharSequence) =
            if (email.startsWith('<') && email.endsWith('>')) containsSubstring(email)
            else containsSubstring("<$email>")

        @JvmStatic
        fun byEmail(email: CharSequence) = or(exactMatch(email), containsEmailAddress(email))

        @JvmStatic
        fun validUserId(key: OpenPGPCertificate) =
            object : SelectUserId() {
                private val info = PGPainless.getInstance().inspect(key)

                override fun invoke(userId: String): Boolean = info.isUserIdValid(userId)
            }

        @JvmStatic
        fun validUserId(keyRing: PGPKeyRing) =
            object : SelectUserId() {
                private val info = PGPainless.inspectKeyRing(keyRing)

                override fun invoke(userId: String): Boolean = info.isUserIdValid(userId)
            }

        @JvmStatic
        fun and(vararg filters: SelectUserId) =
            object : SelectUserId() {
                override fun invoke(userId: String): Boolean = filters.all { it.invoke(userId) }
            }

        @JvmStatic
        fun or(vararg filters: SelectUserId) =
            object : SelectUserId() {
                override fun invoke(userId: String): Boolean = filters.any { it.invoke(userId) }
            }

        @JvmStatic
        fun not(filter: SelectUserId) =
            object : SelectUserId() {
                override fun invoke(userId: String): Boolean = !filter.invoke(userId)
            }
    }
}
